pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go

```diff
@@ -83,30 +92,39 @@ type serviceImportCacheImpl struct {
 	lister   mcsLister.ServiceImportLister
 }
 
+// onServiceEvent is called when the controller receives an event for the kube Service (i.e. cluster.local).
+// When this happens, we need to update the state of the associated synthetic MCS service.
 func (ic *serviceImportCacheImpl) onServiceEvent(svc *model.Service, event model.Event) {
 	if strings.HasSuffix(svc.Hostname.String(), mcsDomainSuffix) {
 		// Ignore events for MCS services that were triggered by this controller.
 		return
 	}
 
-	vips := ic.imports.GetClusterSetIPs(namespacedNameForService(svc))
-	mcsService := ic.newMCSService(svc, vips)
+	namespacedName := namespacedNameForService(svc)
+
+	// Lookup the previous MCS service if there was one.
+	mcsHost := serviceClusterSetLocalHostname(namespacedName)
+	prevMcsService := ic.GetService(mcsHost)
+
+	// Get the ClusterSet VIPs for this service in this cluster. Will only be populated if the
+	// service has a ServiceImport in this cluster.
+	vips := ic.imports.GetClusterSetIPs(namespacedName)
 
-	exists := ic.GetService(mcsService.Hostname) != nil
 	if event == model.EventDelete || len(vips) == 0 {
-		if exists {
+		if prevMcsService != nil {
 			// There are no vips in this cluster. Just delete the MCS service now.
-			ic.deleteService(mcsService)
+			ic.deleteService(prevMcsService)
 		}
 		return
 	}
 
-	if exists {
+	if prevMcsService != nil {
 		event = model.EventUpdate
 	} else {
 		event = model.EventAdd
 	}
 
+	mcsService := ic.genMCSService(svc, mcsHost, vips)
 	ic.addOrUpdateService(nil, mcsService, event)
```

Comment:
I think there is a bug here, or I am misunderstanding the flow:

* Cluster 1 creates Service `foo` with port 80. We trigger this function and store the `model.Service` with port 80
* Cluster 2 creates Service `foo` with port 8080. We trigger this function and _overwrite_ the `model.Service` with port 8080

Reply:
So this is actually something I discussed with @stevenctl yesterday.

Service merging assumes sameness across clusters. If a service is configured differently across clusters, I suspect multi-cluster is just generally broken.